stratum/hal/lib/np4intel/np4_chassis_manager.cc

```diff
@@ -0,0 +1,427 @@
+  return absl::WrapUnique(new NP4ChassisManager(phal_interface));
+}
+
+void NP4ChassisManager::SendPortOperStateGnmiEvent(
+    uint64 node_id, uint32 port_id, PortState new_state) {
+  absl::ReaderMutexLock l(&gnmi_event_lock_);
+  if (!gnmi_event_writer_) return;
+  // Allocate and initialize a PortOperStateChangedEvent event and pass it to
+  // the gNMI publisher using the gNMI event notification channel.
+  // The GnmiEventPtr is a smart pointer (shared_ptr<>) and it takes care of
+  // the memory allocated to this event object once the event is handled by
+  // the GnmiPublisher.
+  if (!gnmi_event_writer_->Write(GnmiEventPtr(
+          new PortOperStateChangedEvent(node_id, port_id, new_state)))) {
+    // Remove WriterInterface if it is no longer operational.
+    gnmi_event_writer_.reset();
+  }
+}
+
+void NP4ChassisManager::ReadPortStatusChangeEvents() {
+  PortStatusChangeEvent event;
+  while (true) {
+    // port_status_change_event_reader_ does not need to be protected by a mutex
+    // because this thread is the only one accessing it. It is assigned in
+    // RegisterEventWriters and then left untouched until UnregisterEventWriters
+    // is called. UnregisterEventWriters joins this thread before resetting the
+    // reader.
+    int code = port_status_change_event_reader_->Read(
+        &event, absl::InfiniteDuration()).error_code();
+    // Exit if the Channel is closed.
+    if (code == ERR_CANCELLED) break;
+    // Read should never timeout.
+    if (code == ERR_ENTRY_NOT_FOUND) {
+      LOG(ERROR) << "Read with infinite timeout failed with ENTRY_NOT_FOUND.";
+      continue;
+    }
+    // Handle received message.
+    {
+      absl::WriterMutexLock l(&chassis_lock);
+      auto* port_id_to_port_state =
+          gtl::FindOrNull(node_id_to_port_id_to_port_state_, event.node_id);
+      if (port_id_to_port_state == nullptr) {
+        LOG(ERROR) << "Node " << event.node_id
+                   << " is not configured or not known.";
+      }
+      auto* port_state_ptr =
+          gtl::FindOrNull(*port_id_to_port_state, event.port_id);
+      if (port_state_ptr == nullptr) {
+        LOG(ERROR) << "Port " << event.port_id
+                   << " is not configured or not known for node "
+                   << event.node_id << ".";
+      }
+      LOG(INFO) << "State of port " << event.port_id << " in node "
+                << event.node_id << ": " << PrintPortState(event.state) << ".";
+      *port_state_ptr = event.state;
+      SendPortOperStateGnmiEvent(event.node_id, event.port_id, event.state);
+    }
+  }
+}
+
+::util::Status NP4ChassisManager::RegisterEventWriters() {
+  if (initialized_) {
+    return MAKE_ERROR(ERR_INTERNAL)
+           << "RegisterEventWriters() can be called only before the class is "
+           << "initialized.";
+  }
+
+  port_status_change_event_channel_ = Channel<PortStatusChangeEvent>::Create(
+      kMaxPortStatusChangeEventDepth);
+
+  port_status_change_event_writer_ =
```

Comment:
warning: writing variable 'port_status_change_event_writer_' requires holding mutex 'port_status_change_event_writer_lock_' exclusively [-Wthread-safety-analysis]
  port_status_change_event_writer_ =
  ^

Reply:
done.  Note: that bmv2 code has the same issue will need to be fixed